README.md
| LINTING            | UNIT          |   INTEGRATION          |  NPM SECURITY       |
|-------------------|-------------------|-------------------|-------------------|
| [![Build1][1]][5] | [![Build2][2]][5] |  [![Build3][3]][5] | [![Build4][4]][5] |

[1]: https://travis-matrix-badges.herokuapp.com/repos/Kinvey/kinvey-cli/branches/master/1
[2]: https://travis-matrix-badges.herokuapp.com/repos/Kinvey/kinvey-cli/branches/master/2
[3]: https://travis-matrix-badges.herokuapp.com/repos/Kinvey/kinvey-cli/branches/master/3
[4]: https://travis-matrix-badges.herokuapp.com/repos/Kinvey/kinvey-cli/branches/master/4
[5]: https://travis-ci.org/Kinvey/kinvey-cli




# Kinvey CLI

Kinvey CLI is a utility for managing various aspects of your Kinvey account from the command line. Its features include:

* Deploying and managing FlexServices running on the Kinvey FlexService Runtime
* Deploying and managing websites
* Creating, listing, and deleting applications
* Creating, listing, and deleting app environments
* Creating, listing, and deleting collections

Contents:

- [Installation](#installation)
- [Usage](#usage)
- [Commands](#commands)
- [Global Options](#global-options)
- [Environment Variables](#environment-variables)
- [Getting Help](#getting-help)
- [Getting Started](#getting-started)
- [Managing Profiles](#managing-profiles)
- [Authenticating One-time Commands](#authenticating-one-time-commands)
- [Precedence of Configuration Options](#precedence-of-configuration-options)
- [Output Format](#output-format)
- [Flex Runtime Version Selection](#flex-runtime-version-selection)
- [Proxy Settings](#proxy-settings)
- [Configuration management](#configuration-management)
- [Troubleshooting](#troubleshooting)
- [Changelog](#changelog)
- [License](#license)



## Installation

Kinvey CLI is distributed as an NPM package. After you install NPM, run the following command to download and install Kinvey CLI.

    npm install -g kinvey-cli

## Usage

    kinvey <command> [args] [options]

## Commands

* `init`

   Prompts you to provide account credentials and instance ID and creates a new working profile for you based on the information you provided. Command-line options that specify the same data are ignored.

* `profile create <name> [profile information]`

   Creates a profile with the specified name. You can specify the profile information either at the command line as arguments or as environment variables. In the presence of command line argument, any values specified through environment variables are ignored.

* `profile list`

   Lists all existing profiles. Profiles are saved under the user home.
   
* `profile login [name]`

    Re-authenticates a specified profile. If you omit the profile name, the active profile is used. Prompts for password and for two-factor authentication token if needed.

* `profile show [name]`

   Shows detailed information about the specified profile. If you omit the profile name, information about the active profile is shown.

* `profile use <name>`

   Sets an active profile.
    
* `profile delete [name]`

   Deletes the specified profile or the active one if you don't specify a profile name.

* `org apply`

    Applies org configuration file to the specified org or to the active one.
    
    * `--org <organization>`
               
        Specifies a Kinvey organization by ID or name.
        
    * `--file <file path>`
        
        Specifies path to an organization configuration file. Required.

* `org list`

    Lists all existing organizations within your Kinvey account.
    
* `org show`

    Shows detailed information about the specified organization or about the active one if you don't specify an organization. You can specify an organization by ID or name.

    * `--org <organization>`
           
        Specifies a Kinvey organization by ID or name.
          
* `org use <org>`

    Sets the specified organization as active. You can specify an organization by ID or name.
    
* `org export`
    
    Exports to a file the specified org or the active one. The organization services and the organization apps along with their services are exported in separate configuration files in the respective 'applications' and 'services' directories.
    
    * `--org <organization>`
               
        Specifies a Kinvey organization by ID or name.
        
    * `--file <file path>`
            
            Specifies path to file. Required.
    
* `app create <name>`

    Creates an application in the specified organization or in the active one.
    
    * `--org <organization>`
               
        Specifies a Kinvey organization by ID or name.
        
    * `--file <file path>`
            
            Specifies path to an application configuration file.

* `app apply`
    
    Applies app configuration file to the specified app or to the active one.
    
    * `--app <application>`
               
        Specifies a Kinvey app by ID or name.
        
    * `--file <file path>`
            
        Specifies path to an application configuration file. Required.
    
* `app list`

    Lists all existing applications within your Kinvey account.
    
* `app show`

    Shows detailed information about the specified application or about the active one if you don't specify an application.
    
    * `--app <application>`
           
        Specifies a Kinvey app by ID or name.

* `app use <app>`

    Sets the specified application as active. You can specify an application by ID or name.
    
* `app export`

    Exports to a file the specified app or the active one. Also exports the app environments and the app services in separate files in the corresponding 'environments' and 'services' directories

    * `--app <application>`
               
        Specifies a Kinvey app by ID or name.
        
    * `--file <file path>`
            
        Specifies path to file. Required.
    
* `app delete`

    Deletes the specified application or the active one if you don't specify an application. You will be prompted for confirmation unless you set the `--no-prompt` flag.
    
    * `--app <application>`
               
        Specifies a Kinvey app by ID or name.
              
    * `--no-prompt`
        
        Do not ask for confirmation.

* `appenv create <name>`

    Creates an environment within the active application. To use a different application, specify it using `--app`.
    
    * `--app <application>`
       
        Specifies a Kinvey app by ID or name.
        
    * `--file <file path>`
            
        Specifies path to an app environment configuration file.
      
* `appenv apply`

    Applies configuration (defined in file) to the specified environment or to the active one if you don't specify an environment. You can specify an environment by ID or name. By default, the command searches inside the active application but you can specify a different application using `--app`.
    
    * `--app <application>`
           
        Specifies a Kinvey app by ID or name.
        
    * `--env <environment>`
              
        Specifies a Kinvey app environment by ID or name.
        
    * `--file <file path>`
            
        Specifies path to an app environment configuration file. Required.
    
* `appenv show`

    Shows detailed information about the specified environment or about the active one if you don't specify an environment. By default, the command searches inside the active application but you can specify a different application using `--app`.
    
    * `--env <environment>`
          
        Specifies a Kinvey app environment by ID or name.
    
    * `--app <application>`
       
        Specifies a Kinvey app by ID or name.

* `appenv use <env>`

    Sets the specified environment as active. By default, the command searches inside the active application but you can specify a different application using `--app`.
    
    * `--app <application>`
       
        Specifies a Kinvey app by ID or name.

* `appenv export`
    
    Exports to a file the specified app environment or the active one.
    
    * `--env <environment>`
              
        Specifies a Kinvey app environment by ID or name.
        
    * `--app <application>`
           
        Specifies a Kinvey app by ID or name.
        
    * `--file <file path>`
            
        Specifies path to file. Required.
    
* `appenv delete`

    Deletes the specified environment or the active one if you don't specify an environment. By default, the command searches inside the active application but you can specify a different application using `--app`. You will be prompted for confirmation unless you set the `--no-prompt` flag.
    
    * `--env <environment>`
          
        Specifies a Kinvey app environment by ID or name.
    
    * `--app <application>`
       
      Specifies a Kinvey app by ID or name.
      
    * `--no-prompt`
        
        Do not ask for confirmation.

* `coll create <name>`

    Creates a collection within the active application and environment. You can specify another pair of application and environment using the `--app` and `--env` options.
    
    * `--app <application>`
       
        Specifies a Kinvey app by ID or name. Requires `--env`.
    
    * `--env <environment>`
      
        Specifies a Kinvey app environment by ID or name.
    
* `coll list`

    Lists all existing collections within the active application and environment. You can specify another pair of application and environment using the `--app` and `--env` options.

    * `--app <application>`
       
        Specifies a Kinvey app by ID or name. Requires `--env`.
    
    * `--env <environment>`
      
        Specifies a Kinvey app environment by ID or name.

* `coll delete <coll>`
    
    Deletes a collection by name within the active application and environment. You can specify another pair of application and environment using the `--app` and `--env` options. You will be prompted for confirmation unless you set the `--no-prompt` flag.

    * `--app <application>`
       
        Specifies a Kinvey app by ID or name. Requires `--env`.
    
    * `--env <environment>`
      
        Specifies a Kinvey app environment by ID or name.
      
    * `--no-prompt`
        
        Do not ask for confirmation.

* `service create <name>`

    Creates a service within an organization (either `--org <organization>`must be specified or active org must be set).
      
    * `--org <organization>`
                
      Specifies a Kinvey organization by ID or name.
      
    * `--file <file path>`
            
        Specifies path to a service configuration file. Required.

* `service apply`

    Applies a service configuration file to the specified service.
    
    * `--service <service>`
        
        Specifies service ID.
        
    * `--file <file path>`
            
        Specifies path to a service configuration file. Required.

* `service export`

    Exports the specified service to a file.

CHANGELOG.md
# Changelog

## 6.2.4 (2023-06-01)
* Only bump version

## 6.2.3 (2023-02-24)
* Fix a security issue by updating `moment`

## 6.2.1 (2021-11-09)
* Add password reset URL support for appEnv config files

## 6.2.0 (2021-06-09)
* Add support for login hooks

## 6.1.3 (2021-03-05)
* Improve error message for iOS push settings in config management

## 6.1.2 (2021-02-16)
* Modify collection hooks sequentially
* Add collection hooks: onPreInsert, onPreUpdate, onPostInsert, onPostUpdate

## 6.1.1 (2020-10-30)
* Do not update the appEnv when no settings are supplied on `kinvey appenv apply`
* Fix appEnv ID check

## 6.1.0 (2020-09-14)
* Support login with external identity providers

## 6.0.0 (2020-07-02)
* Modify commands to accept only `--org <org>` (`--app <app>` is no longer valid):
    * `flex init`
    * `flex create`
    * `flex list`
    * `service create`
    * `website create`
    * `app create`
* Remove `flex job` command
* Update API version to 4
* Update min node version to 12.16.1

## 5.0.0 (2019-11-28)
* Move `file` from positional argument to option
* Rename `push` commands to `apply`
* Allow creating app, app env or service with a name that contains only digits
* Do not export the `serviceEnvironment` field on the serviceObject during `service export`
* Fetch all orgs that a user can read
* Do not export services of type custom
* Remove support for SAP services as they are deprecated
* Allow `insertMany` method for services
* Improve profile description 

## 5.0.0-alpha.1 (2019-10-11)
* Fix BAAS host slashes when backend runs on localhost
* Add `org push` to modify organizations from config file
* Fix config management by adjusting to services endpoints changes
* Improve response error handling

## 5.0.0-alpha.0 (2018-12-14)
* Config management related logic uses v3 API
* Support service environments in service config files
* Support runtime and environment variables in flex service config files
* Fix 'app push' to update properly env metadata

## 4.5.2 (2019-09-19)
* Validate files size when deploying websites

## 4.5.1 (2019-08-01)
* Output error if identifier (e.g. `--app`) matches more than one entity
* Update `lodash.merge` to `4.6.2`

## 4.5.0 (2019-05-16)
* `website deploy` command improvements:
    * verify files for index and/or error page are present
    * add `--force` flag to force deploy
* Add flex runtime version `node12`

## 4.4.0 (2019-04-25)
* Add websites support:
    * `website create <name>`
    * `website show`
    * `website list`
    * `website deploy`
    * `website publish`
    * `website status`
    * `website unpublish`
    * `website delete`

## 4.1.3 (2019-03-20)
* Show any 'status' value in all caps (`flex status`)

## 4.1.2 (2019-03-05)
* Suggest command when token is expired
* Use backend error when HTTP status code is 401
* Apply Kinvey lint rules and fix lint errors

## 4.1.1 (2018-11-29)
* Preserve existing env vars on `flex deploy --set-vars <env variables>` 

## 4.1.0 (2018-11-27)
* Migrate to v3 API
* Add service environments functionality in flex namespace

## 4.1.0-alpha.3
* Add `org export` to export organizations
* Ensure `export` creates non-existing directories
* App functionality to create/modify app from config file
* Add `app export` command to export applications
* Support rapid data and flex service types for `service export`
* Specify entity identifiers as options

## 4.1.0-alpha.2
* Add `service export` command to export a flex service to config file
* Output logs properly when log message is an object
* Prompt for 2FA token on 'profile login' if required
* Rearrange logs

## 4.1.0-alpha.1
* Add support for groups in env config file
* Fix env export issues

## 4.1.0-alpha.0
* Add functionality to create/modify env from config file
* Add functionality to export env to config file

## 4.0.1 (2018-11-21)
* Add runtime environment support for internal flex services

## 4.0.0 (2018-11-08)
* Add commands:
    * organizations-related: `org list`, `org show`, `org use`
    * applications-related: `app create`, `app list`, `app show`, `app use`, `app delete`
    * app environments-related: `appenv create`, `appenv list`, `appenv show`, `appenv use`, `appenv delete`
    * collections-related: `coll create`, `coll list`, `coll delete`
* Add `org` option to `app create`
* Add `flex create` command to create internal flex services
* Add `flex clear` command to clear project settings
* Add environment variables for internal flex services
* Repurpose the `flex delete` command to delete internal flex services
* Ask for confirmation on app/appenv/coll/flex delete
* Fail on unsupported hyphenated option or flag
* Update production dependencies (moment.js, request.js)
* Update dev dependencies
* Show help on `kinvey <namespace>`
* Fix `flex deploy` to work with local setup
* Fix `profile login` to update only token

## 3.1.2
* Enhance `flex status` output

## 3.1.1
* Prompt for 2FA token on 'profile login' if required
* Rearrange logs
* Output logs properly when log message is an object

## 3.1.0
* Ask for profile override if profile already exists on 'kinvey init'
* Setup Travis
* Use active profile if no other specified on 'profile delete'
* Add 'profile login'
* Add 2fa option

## 3.0.0

## 3.0.0-beta.4
* Apply Kinvey es6 style guide
* Fix 'flex init' when no config file

## 3.0.0-beta.3
* Refactor command handlers
* Fail on too many args
* Replace host with instanceId
* Configure flex project for many instances
* Provide shorthand option for number

## 3.0.0-beta.2

lib/Utils.js
/**
 * Copyright (c) 2018, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 */

const fs = require('fs');
const os = require('os');
const path = require('path');

const AESjs = require('aes-js');
const async = require('async');
const chalk = require('chalk');
const validUrl = require('valid-url');
const isemail = require('isemail');
const uuidV4 = require('uuid-v4');
const isempty = require('lodash.isempty');
const lodashGet = require('lodash.get');
const lodashCloneDeep = require('lodash.clonedeep');
const lodashPick = require('lodash.pick');
const lodashOmitBy = require('lodash.omitby');
const lodashSortBy = require('lodash.sortby');
const lodashPickBy = require('lodash.pickby');
const logger = require('./logger.js');
const moment = require('moment');
const open = require('open');
const { ActiveItemTypes, Errors, PromptMessages } = require('./Constants');
const KinveyError = require('./KinveyError');

const Utils = {};
Utils.formatHost = function formatHost(host) {
  if (Utils.isNullOrUndefined(host) || Utils.isStringWhitespace(host)) {
    return 'https://manage.kinvey.com/';
  }

  let validHost = validUrl.isHttpUri(host);
  if (validHost) {
    if (validHost.slice(-1) !== '/') {
      validHost = `${validHost}/`;
    }
    return validHost;
  }

  validHost = validUrl.isHttpsUri(host);
  if (validHost) {
    if (validHost.slice(-1) !== '/') {
      validHost = `${validHost}/`;
    }
    return validHost;
  }

  return `https://${host}-manage.kinvey.com/`;
};

Utils.formatList = function formatList(list, name = 'name') {
  const result = list.map(el => ({
    name: el[name],
    value: el
  }));
  result.sort((x, y) => {
    if (x[name].toLowerCase() < y[name].toLowerCase()) return -1;
    return 1;
  });
  return result;
};

Utils.formatHostList = function formatHostList(list) {
  const result = list.map(el => ({
    name: el,
    value: el
  }));
  result.unshift({
    name: 'all hosts',
    value: null
  });
  return result;
};

Utils.readFile = function readFile(file, cb) {
  logger.debug('Reading contents from file %s', chalk.cyan(file));
  return fs.readFile(file, cb);
};

Utils.readJSON = function readJSON(file, cb) {
  logger.debug('Reading JSON from file %s', chalk.cyan(file));
  return async.waterfall([
    next => Utils.readFile(file, next),
    (data, next) => {
      let json = null;
      try {
        logger.debug('Parsing JSON from file: %s', chalk.cyan(file));
        json = JSON.parse(data);
      } catch (error) {
        logger.warn('Invalid JSON in file %s', chalk.cyan(file));
        return next(error);
      }
      return next(null, json);
    }
  ], cb);
};

/**
 * Reads a file synchronously and returns the data in JSON. Exceptions are not caught.
 * @param {string} filePath
 */
Utils.readJSONSync = function readJSONSync(filePath) {
  const rawData = fs.readFileSync(filePath);
  const jsonData = JSON.parse(rawData);
  return jsonData;
};

Utils.writeFile = function writeFile(file, contents, cb) {
  logger.debug('Writing contents to file %s', chalk.cyan(file));
  return fs.writeFile(file, contents, cb);
};

Utils.writeJSON = function writeJSON({ file, data, createParentDir, space }, done) {
  async.series([
    (next) => {
      if (!createParentDir) {
        return setImmediate(next);
      }

      Utils.mkdirp(file, next);
    },
    (next) => {
      const contents = JSON.stringify(data, null, space);
      Utils.writeFile(file, contents, next);
    }
  ], done);
};

Utils.writeConfigFile = function writeConfigFile(filePath, data, done) {
  Utils.writeJSON({ file: filePath, data, createParentDir: true, space: 4 }, (err) => {
    done(err, data);
  });
};

Utils.isValidMFAToken = function isValidMFAToken(value) {
  return (/^\d{6}$/.test(value));
};

Utils.isValidEmail = function isValidEmail(value) {
  return !Utils.isNullOrUndefined(value) && isemail(value);
};

Utils.isNullOrUndefined = function isNullOrUndefined(value) {
  return value == null;
};

Utils.isEmpty = function isEmpty(value) {
  return isempty(value);
};

Utils.isValidTimestamp = function isValidTimestamp(ts) {
  return moment(ts, moment.ISO_8601, true).isValid();
};

Utils.isValidNonZeroInteger = function isValidNonZeroInteger(number) {
  if (number === 0 || number === '0') return false;
  return /^\d+$/.test(number);
};

Utils.validateEmail = function validateEmail(value) {
  if (Utils.isValidEmail(value)) {
    return true;
  }

  return PromptMessages.INVALID_EMAIL_ADDRESS;
};

Utils.isStringWhitespace = function isStringWhitespace(value) {
  const trimmedValue = value.trim();
  return trimmedValue.length < 1;
};

Utils.askForValue = function askForValue(value) {
  if (Utils.isNullOrUndefined(value)) {
    return true;
  }

  return false;
};

Utils.validateString = function validateString(value) {
  if (!Utils.isStringWhitespace(value)) {
    return true;
  }

  return PromptMessages.INVALID_STRING;
};

Utils.validateMFAToken = function validateMFAToken(value) {
  if (Utils.isValidMFAToken(value)) {
    return true;
  }

  return PromptMessages.INVALID_MFA_TOKEN;
};

Utils.Endpoints = {
  getIdPartFromId: id => (Utils.isNullOrUndefined(id) ? '' : `/${id}`),
  version: schemaVersion => `v${schemaVersion}`,
  session: () => 'session',
  apps: (schemaVersion, appId) => `${Utils.Endpoints.version(schemaVersion)}/apps${Utils.Endpoints.getIdPartFromId(appId)}`,
  envs: (schemaVersion, envId) => `${Utils.Endpoints.version(schemaVersion)}/environments${Utils.Endpoints.getIdPartFromId(envId)}`,
  push: (schemaVersion, envId) => `${Utils.Endpoints.envs(schemaVersion, envId)}/push`,
  envsByAppId: (schemaVersion, appId) => `${Utils.Endpoints.apps(schemaVersion, appId)}/environments`,
  collections: (schemaVersion, envId, collName) => `${Utils.Endpoints.envs(schemaVersion, envId)}/collections${Utils.Endpoints.getIdPartFromId(collName)}`,
  commonCode: (schemaVersion, envId, name) => `${Utils.Endpoints.envs(schemaVersion, envId)}/business-logic/common${Utils.Endpoints.getIdPartFromId(name)}`,
  endpoints: (schemaVersion, envId, name) => `${Utils.Endpoints.envs(schemaVersion, envId)}/business-logic/endpoints${Utils.Endpoints.getIdPartFromId(name)}`,
  hooks: (schemaVersion, envId, collName, hookName) => `${Utils.Endpoints.envs(schemaVersion, envId)}/business-logic/collections/${collName}${Utils.Endpoints.getIdPartFromId(hookName)}`,
  collectionHooks: (schemaVersion, envId, collectionName) => {
    const namePart = collectionName ? `/${collectionName}` : '';
    return `${Utils.Endpoints.envs(schemaVersion, envId)}/business-logic/collections${namePart}`;
  },
  orgs: (schemaVersion, orgId) => `${Utils.Endpoints.version(schemaVersion)}/organizations${Utils.Endpoints.getIdPartFromId(orgId)}`,
  appsByOrg: (schemaVersion, orgId) => `${Utils.Endpoints.orgs(schemaVersion, orgId)}/apps`,
  jobs: (schemaVersion, id) => `${Utils.Endpoints.version(schemaVersion)}/jobs${Utils.Endpoints.getIdPartFromId(id)}`,
  services: (schemaVersion, serviceId) => `${Utils.Endpoints.version(schemaVersion)}/services${Utils.Endpoints.getIdPartFromId(serviceId)}`,
  serviceEnvs: (schemaVersion, serviceId, svcEnvId) => `${Utils.Endpoints.services(schemaVersion, serviceId)}/environments${Utils.Endpoints.getIdPartFromId(svcEnvId)}`,
  serviceStatus: (schemaVersion, serviceId, svcEnvId) => `${Utils.Endpoints.serviceEnvs(schemaVersion, serviceId, svcEnvId)}/status`,
  serviceLogs: (schemaVersion, serviceId, svcEnvId) => `${Utils.Endpoints.serviceEnvs(schemaVersion, serviceId, svcEnvId)}/logs`,
  sites: (schemaVersion, siteId) => `${Utils.Endpoints.version(schemaVersion)}/sites${Utils.Endpoints.getIdPartFromId(siteId)}`,
  siteEnvs: (schemaVersion, siteId, siteEnvId) => `${Utils.Endpoints.sites(schemaVersion, siteId)}/environments${Utils.Endpoints.getIdPartFromId(siteEnvId)}`,
  siteDeploy: (schemaVersion, siteId, siteEnvId) => `${Utils.Endpoints.siteEnvs(schemaVersion, siteId, siteEnvId)}/files`,
  sitePublish: (schemaVersion, siteId) => `${Utils.Endpoints.sites(schemaVersion, siteId)}/publish`,
  siteUnpublish: (schemaVersion, siteId) => `${Utils.Endpoints.sites(schemaVersion, siteId)}/unpublish`,
  siteStatus: (schemaVersion, siteId) => `${Utils.Endpoints.sites(schemaVersion, siteId)}/status`,
  identityProviders: schemaVersion => `${Utils.Endpoints.version(schemaVersion)}/identity-providers`,
  identityProvidersAuthenticate: (schemaVersion, providerId) => `${Utils.Endpoints.version(schemaVersion)}/identity-providers/${providerId}/authenticate`,
  user: schemaVersion => `${Utils.Endpoints.version(schemaVersion)}/user`
};

Utils.getCommandNameFromOptions = function getCommandNameFromOptions(options) {
  let name;
  if (options._ && options._.length) {
    name = options._.join(' ');
  }

  return name;
};

Utils.getErrorFromRequestError = function getErrorFromRequestError(err) {
  let errResult;
  const errCode = err.code;
  if (errCode === 'ENOTFOUND') {
    errResult = new KinveyError(Errors.InvalidConfigUrl);
  } else if (errCode === 'ETIMEDOUT') {
    errResult = new KinveyError(Errors.RequestTimedOut);
  } else if (errCode === 'ECONNRESET') {
    errResult = new KinveyError(Errors.ConnectionReset);
  } else if (errCode === 'ECONNREFUSED') {
    errResult = new KinveyError(errCode, `Connection refused at ${err.address}`);
  } else {
    errResult = err;
  }

  return errResult;
};

Utils.isMFATokenError = function isMFATokenError(err) {
  return err && err.name === 'InvalidTwoFactorAuth';
};

Utils.findAndSortInternalServices = function findAndSortInternalServices(services) {
  const result = services.filter(el => el.type === 'internal');
  result.sort((x, y) => {
    if (x.name.toLowerCase() < y.name.toLowerCase()) {
      return -1;
    }

    return 1;
  });

  return result;
};

Utils.isArtifact = function isArtifact(artifacts, base, filepath) {
  const relative = path.normalize(path.relative(base, filepath));
  for (let i = 0; i < artifacts.length; i += 1) {
    const pattern = artifacts[i];
    if (relative.indexOf(pattern) === 0 || (`${relative}/`) === pattern) return true;
  }
  return false;
};

Utils.getDeviceInformation = function getDeviceInformation(cliVersion) {
  const info = `kinvey-cli/${cliVersion} ${os.platform()} ${os.release()}`;
  return info;
};

Utils.mkdirp = function mkdirp(pathToFile, done) {
  try {
    const filepath = path.resolve(pathToFile);
    filepath.split(path.sep).slice().reduce((prev, curr, i) => {
      if (prev && !fs.existsSync(prev)) {
        fs.mkdirSync(prev);
      }
      return prev + path.sep + curr;
    });
    done();
  } catch (err) {
    done(err);
  }
};

/**
 * Asserts whether a value is of UUID v4 format. Values without dashes are accepted, too.
 * Returns false if value is null or undefined.
 * @param {String} value
 * @returns {*}
 */
Utils.isUUID = function isUUID(value) {
  if (Utils.isNullOrUndefined(value)) {
    return false;
  }

  const dash = '-';
  const lengthWithDashes = 36;
  const lengthValue = value.length;
  const couldBeRegularUUUID = value.includes(dash) && lengthValue === lengthWithDashes;
  if (couldBeRegularUUUID) {
    return uuidV4.isUUID(value);
  }

  const lengthWithoutDashes = 32;
  const couldBeUUIDWithoutDashes = lengthValue === lengthWithoutDashes;
  if (!couldBeUUIDWithoutDashes) {
    return false;
  }

  const valueWithDashes = `${value.slice(0, 8)}${dash}${value.slice(8, 12)}${dash}${value.slice(12, 16)}${dash}${value.slice(16, 20)}${dash}${value.slice(20)}`;
  return uuidV4.isUUID(valueWithDashes);
};

Utils.isEnvID = function isEnvID(value) {
  if (Utils.isNullOrUndefined(value)) {
    return false;
  }

  // kid_SklZwh7dN
  return value.startsWith('kid_');
};

Utils.validateActiveItemType = function validateActiveItemType(itemType) {
  if (!ActiveItemTypes.includes(itemType)) {
    throw new KinveyError(`Invalid item type: ${itemType}.`);
  }
};

/**
 * Returns error for ItemNotSpecified.
 * @param {Constants.EntityType} entityType
 * @returns {KinveyError}
 */
Utils.getItemError = function getItemError(entityType) {
  const msg = `No ${entityType} identifier is specified and active ${entityType} is not set.`;
  return new KinveyError(Errors.ItemNotSpecified.NAME, msg);
};

Utils.getConfigTypeError = function getConfigTypeError(type) {
  const msg = `Unrecognized config type: ${type}`;
  return new KinveyError('ValidationError', msg);
};

Utils.isEntityError = function isEntityError(err) {
  if (Utils.isNullOrUndefined(err)) {
    return false;
  }

  return err.name === Errors.NoEntityFound.NAME || err.name === Errors.TooManyEntitiesFound.NAME;
};

/**
 * Returns error for generic entity errors with custom message.
 * @param {Constants.EntityType} entityType
 * @param {String} identifier
 * @param {String} errName Error name. Must be either 'NotFound' or 'TooManyFound'.
 * @returns {KinveyError}
 */
Utils.getCustomEntityError = function getCustomEntityError(entityType, identifier, errName) {
  if (errName === Errors.NoEntityFound.NAME) {
    return Utils.getCustomNotFoundError(entityType, identifier);
  }

  if (errName === Errors.TooManyEntitiesFound.NAME) {
    const lastPart = identifier ? ` with identifier '${identifier}'.` : '.';
    const msg = `Found too many ${entityType}s${lastPart}`;
    return new KinveyError(Errors.TooManyEntitiesFound.NAME, msg);
  }

  return new KinveyError('UnknownError', `Failed to construct proper error for error name: ${errName}`);
};

Utils.getCustomNotFoundError = function getCustomNotFoundError(entityType, identifier) {
  const lastPart = identifier ? ` with identifier '${identifier}'.` : '.';
  const msg = `Could not find ${entityType}${lastPart}`;
  return new KinveyError(Errors.NoEntityFound.NAME, msg);
};

/**
 * Returns original error or an error with custom message (if initial error is a generic entity error).
 * @param {Error} err
 * @param {Constants.EntityType} entityType
 * @param {String} identifier
 * @returns {Error}
 */
Utils.getTransformedError = function getTransformedError(err, entityType, identifier) {
  if (!Utils.isEntityError(err)) {
    return err;
  }

  return Utils.getCustomEntityError(entityType, identifier, err.name);
};

Utils.getValueFromObject = function getValueFromObject(obj, path, defaultValue = 'Not set') {
  return lodashGet(obj, path, defaultValue);
};

Utils.mapFromSource = function mapFromSource(mapping, originalSource) {
  const mappingKeys = Object.keys(mapping);
  const sourceIsArr = Array.isArray(originalSource);
  let source;
  if (!sourceIsArr) {
    source = [originalSource];
  } else {
    source = originalSource;
  }

  const result = source.map((item) => {
    const targetItem = {};

    mappingKeys.forEach((k) => {
      let targetKeyValue;
      const valueOfMappingKey = mapping[k];
      if (typeof valueOfMappingKey === 'function') {
        targetKeyValue = valueOfMappingKey(item);
      } else {
        targetKeyValue = Utils.getValueFromObject(item, mapping[k]);
      }

      targetItem[k] = targetKeyValue;
    });

    return targetItem;
  });

  return sourceIsArr ? result : result[0];
};

Utils.getObjectByOmitting = function getObjectByOmitting(source, propsToOmit) {
  const result = {};
  const keys = Object.keys(source);
  keys.forEach((k) => {
    if (!propsToOmit.includes(k)) {
      result[k] = lodashCloneDeep(source[k]);
    }
  });

  return result;
};

Utils.getObjectByPicking = function getObjectByPicking(source, propsToPick) {
  return lodashPick(source, propsToPick);
};

/**
 * Returns an object from a string of key-value pairs where key1[secondaryDelimiter]value[primaryDelimiter] (e.g key1=value1,key2=value2).
 * Throws an error if pattern is not valid.
 * @param {String} dsv Delimiter separated key-values.
 * @param {String} primaryDelimiter Delimiter that separates key-value pairs.
 * @param {String} secondaryDelimiter Delimiter that separates keys from values.
 * @returns {Object}
 */
Utils.getObjectFromDelimiterSeparatedKeyValuePairs = function getObjectFromDelimiterSeparatedKeyValuePairs(dsv, primaryDelimiter = ',', secondaryDelimiter = '=') {
  if (typeof dsv !== 'string') {
    throw new Error('Expected string.');
  }

  const listOfKeyValuePairs = dsv.split(primaryDelimiter);
  if (Utils.isEmpty(listOfKeyValuePairs)) {
    throw new Error('Empty list.');
  }

  const result = {};
  listOfKeyValuePairs.reduce((accumulator, pair) => {
    const delimiterIndex = pair.indexOf(secondaryDelimiter);
    if (delimiterIndex < 0) {
      throw new Error(`No delimiter ('${secondaryDelimiter}') in '${pair}'.`);
    }

    const key = pair.substring(0, delimiterIndex);
    const value = pair.substring(delimiterIndex + 1, pair.length);
    if (key.length < 1 || value.length < 1) {
      throw new Error(`Invalid key-value pair: ${pair}`);
    }

    accumulator[key] = value;
    return accumulator;
  }, result);

  return result;
};

/**
 * Sorts a list of objects.
 * @param {Array} list
 * @param {String} [property] Property to sort by. Defaults to 'name'.
 * @returns {Array.<Object>}
 */
Utils.sortList = function sortList(list, property = 'name') {
  return lodashSortBy(list, x => x[property].toLowerCase());
};

Utils.pickBy = function pickBy(source, predicate) {
  return lodashPickBy(source, predicate);
};

/**
 * Returns a message that contains info about the entity in use.
 * @param {Constants.EntityType} entityType
 * @param identifier
 * @returns {string}
 */
Utils.getUsingEntityMsg = function getUsingEntityMsg(entityType, identifier) {
  return `Using ${entityType}: ${identifier}`;
};

Utils.getDefaultIdentityProvider = function getDefaultIdentityProvider() {
  return {
    id: 'Kinvey',
    name: 'Kinvey'
  };
};

/**
 * Returns secret key. 256-bit key derived from 58 digit key.
 * @returns {String}
 */
Utils.generateSecretKey = function generateSecretKey() {
  let ret = '';
  const length = 58;
  while (ret.length < length) {
    ret += Math.random().toString(16).substring(2);
  }

  const text = ret.substring(0, length);

  const key256 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28,
    29, 30, 31];
  const textBytes = AESjs.utils.utf8.toBytes(text);

  // The counter is optional, and if omitted will begin at 1
  const aesCtr = new AESjs.ModeOfOperation.ctr(key256, new AESjs.Counter(5)); // eslint-disable-line
  const encryptedBytes = aesCtr.encrypt(textBytes);
  const result = AESjs.utils.hex.fromBytes(encryptedBytes);
  return result;
};

/**
 * Removes keys from object that have value of null or undefined. Returns new object.
 * @param {Object} obj
 * @returns {Object}
 */
Utils.stripNullOrUndefined = function stripNullOrUndefined(obj) {
  return lodashOmitBy(obj, Utils.isNullOrUndefined);
};

Utils.openBrowser = function openBrowser(url) {
  return open(url);
};

module.exports = Utils;
